spec/javascripts/miq_application_spec.js

```diff
@@ -497,5 +497,65 @@ describe('miq_application.js', function() {
       });
     });
   });
-});
 
+  describe('User login form', function() {
+    beforeEach(function() {
+      var html = '<div class="form-horizontal" id="login_div">\
+        <div class="form-group">\
+          <label class="col-md-3 control-label">Username</label>\
+          <div class="col-md-9">\
+            <input type="text" name="user_name" id="user_name" class="form-control" placeholder="Username" onkeypress="if (miqEnterPressed(event)) miqAjaxAuth();">\
+          </div>\
+        </div>\
+        <div class="form-group">\
+          <label class="col-md-3 control-label">Password</label>\
+          <div class="col-md-9">\
+            <input type="password" name="user_password" id="user_password" onkeypress="if (miqEnterPressed(event)) miqAjaxAuth();" autocomplete="off" placeholder="Password" class="form-control">\
+          </div>\
+        </div>\
+        <div class="form-group">\
+          <div class="col-xs-8 col-md-offset-3 col-md-6">\
+            <div id="back_button" style="display: none">\
+              <a data-method="post" title="Back" data-remote="true" href="/dashboard/authenticate?button=back">Back</a>\
+            </div>\
+          </div>\
+        </div>\
+        <div class="col-xs-4 col-md-3 submit">\
+          <a id="login" class="btn btn-primary" alt="Log in" title="Log in" onclick="miqAjaxAuth("/dashboard/authenticate?button=login"); return false;" href="">Log in</a>\
+          <a id="sso_login" class="btn btn-primary" alt="SSO Log in" title="SSO Log in" style="display: none;" onclick="miqAjaxAuthSso("/dashboard/kerberos_authenticate?button=sso_login"); return false;" href="">SSO Log in</a>\
+        </div>\
+      </div>';
+
+      setFixtures(html);
+    });
+    context('failed login', function() {
+      var deferred;
+
+      beforeEach(module('ManageIQ'));
+      beforeEach(
+        inject(function(_API_, $q) {
+          var API = _API_;
+          deferred = $q.defer();
+
+          // simulate failed authentication api call
+          spyOn(API, 'get').and.callFake(function() {return deferred.reject();});
```

Comment:
Sorry for the late response.. this spec doesn't do anything... (thanks @mzazrivec for noticing :))

Fixed in https://github.com/ManageIQ/manageiq-ui-classic/pull/3355, but just so you know :)

* can't use `inject(API)` because this is not angular code - login is using `vanillaJsAPI`, not the angular service
* can't mock `API.get` because it is not actually used, `vanillaJsAPI.login` is called and does the `get` internally
* any time we use a promise.then/catch, the code is run asynchronously - and asynchronous expects are ignored in synchronous tests - missing a `done` param for the `it` function, jasmine uses the arity of the function to determine whether it's synchronous (arity 0) or asynchronous (arity 1 and you need to call the provided callback when done), so this was a synchronous spec with 0 expects
* having the mock return a fake promise would make sense if we didn't have to wait, but we can't wait on a then.then.then of a promise without getting the new promise - so `miqAjaxAuth` needs to return the right promise so that we can wait on it
* missing spy on `miqClearLoginFields`
* badly escaped quotes-within-quotes-within-apostrophes (but we can't really use the onclick anyway)